### PR TITLE
[SPARK-11124] JsonParser/Generator should be closed for resource recycle

### PR DESCRIPTION
**Pull request by navis**
Some json parsers are not closed. parser in JacksonParser#parseJson, for example.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 1]
```diff
@@ -2153,6 +2153,9 @@ private[spark] object Utils extends Logging {
       conf.getInt("spark.executor.instances", 0) == 0
   }
 
+  def withResource[R <: Closeable, T](resource: R, f: () => T): T = {
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: It's probably best to do it like this
```scala
def withResource[R <: Closeable, T](createResource: => R)(f: R => T): T = {

}
```

then you can call it like
```scala
withResource(factory.createParser(row)) { parser =>
  ...
}
```

Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: What do you think about naming this `tryWithResource` to evoke Java's try-with-resources statement?

Comment (rxin) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: +1

Comment (navis) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: @rxin It's just what I wanted to do (but couldn't). Thanks.
@JoshRosen Good.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 2]
```diff
@@ -251,16 +254,18 @@ case class GetJsonObject(json: Expression, path: Expression)
...
         val buffer = new StringWriter()
         val flattenGenerator = jsonFactory.createGenerator(buffer)
-        flattenGenerator.writeStartArray()
 
         var dirty = 0
-        while (p.nextToken() != END_ARRAY) {
-          // track the number of array elements and only emit an outer array if
-          // we've written more than one element, this matches Hive's behavior
-          dirty += (if (evaluatePath(p, flattenGenerator, nextStyle, xs)) 1 else 0)
+        Utils.tryWithResource(jsonFactory.createGenerator(buffer)) { flattenGenerator =>
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: we should delete the outer flattenGenerator here?

Comment (navis) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Ah, missed that.

**Comment by rxin** [comment]
Thanks - I've merged this.

**Pull request closed without merging**